lib/c-api/src/wasm_c_api/unstable/target_lexicon.rs

```diff
@@ -0,0 +1,318 @@
+//!
+//!         wasm_cpu_features_add(cpu_features, &cpu_feature_name);
+//!
+//!         wasm_name_delete(&cpu_feature_name);
+//!     }
+//!
+//!     assert(cpu_features);
+//!
+//!     // Create the target!
+//!     wasm_target_t* target = wasm_target_new(triple, cpu_features);
+//!     assert(target);
+//!
+//!     wasm_target_delete(target);
+//!
+//!     return 0;
+//! }
+//! #    })
+//! #    .success();
+//! # }
+//! ```
+
+use super::super::types::wasm_name_t;
+use crate::error::CApiError;
+use enumset::EnumSet;
+use std::ffi::CStr;
+use std::slice;
+use std::str::FromStr;
+use wasmer_compiler::{CpuFeature, Target, Triple};
+
+/// Unstable non-standard Wasmer-specific API to represent a triple +
+/// CPU features pair.
+///
+/// # Example
+///
+/// See the module's documentation.
+#[derive(Debug)]
+#[allow(non_camel_case_types)]
+pub struct wasm_target_t {
+    pub(crate) inner: Target,
+}
+
+/// Creates a new [`wasm_target_t`].
+///
+/// It takes ownership of `triple` and `cpu_features`.
+///
+/// # Example
+///
+/// See the module's documentation.
+#[no_mangle]
+pub unsafe extern "C" fn wasm_target_new(
+    triple: Option<Box<wasm_triple_t>>,
+    cpu_features: Option<Box<wasm_cpu_features_t>>,
+) -> Option<Box<wasm_target_t>> {
+    let triple = triple?;
+    let cpu_features = cpu_features?;
+
+    Some(Box::new(wasm_target_t {
+        inner: Target::new(triple.inner.clone(), cpu_features.inner.clone()),
+    }))
+}
+
+/// Delete a [`wasm_target_t`].
+///
+/// # Example
+///
+/// See the module's documentation.
+#[no_mangle]
+pub unsafe extern "C" fn wasm_target_delete(_target: Option<Box<wasm_target_t>>) {}
+
+/// Unstable non-standard Wasmer-specific API to represent a target
+/// “triple”.
+///
+/// Historically such things had three fields, though they have added
+/// additional fields over time.
+///
+/// # Example
+///
+/// ```rust
+/// # use inline_c::assert_c;
+/// # fn main() {
+/// #    (assert_c! {
+/// # #include "tests/wasmer_wasm.h"
+/// #
+/// int main() {
+///     wasm_name_t triple_name;
+///     wasm_name_new_from_string(&triple_name, "x86_64-apple-darwin");
+///
+///     wasm_triple_t* triple = wasm_triple_new(&triple_name);
+///     assert(triple);
+///
+///     wasm_triple_delete(triple);
+///     wasm_name_delete(&triple_name);
+///
+///     return 0;
+/// }
+/// #    })
+/// #    .success();
+/// # }
+/// ```
+///
+/// See also [`wasm_triple_new_from_host`].
+#[allow(non_camel_case_types)]
+pub struct wasm_triple_t {
+    inner: Triple,
+}
+
+/// Create a new [`wasm_triple_t`] based on a triple string.
+///
+/// # Example
+///
+/// See [`wasm_triple_t`] or [`wasm_triple_new_from_host`].
+#[no_mangle]
+pub unsafe extern "C" fn wasm_triple_new(
+    triple: Option<&wasm_name_t>,
+) -> Option<Box<wasm_triple_t>> {
+    let triple = triple?;
+    let triple = c_try!(CStr::from_bytes_with_nul_unchecked(slice::from_raw_parts(
+        triple.data,
+        triple.size + 1
```

Comment:
Why is this correct?

Reply:
🤦 

`CStr` expects a C string, a string that is nul-terminated. It's not the case here. I've updated the code to use `str::from_utf8(slice::from_raw_parts(triple.data, triple.size))`, which is the correct way to use `wasm_name_t` :-).

Thanks!